src/main/java/com/google/api/codegen/util/NameFormatterMixin.java

```diff
@@ -0,0 +1,73 @@
+/* Copyright 2016 Google Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.api.codegen.util;
+
+/**
+ * NameFormatterMixin is an abstract class that implements NameFormatter and
+ * simply forwards calls to another NameFormatter. This allows a child class
+ * to provide the interface of NameFormatter along with additional functionality.
+ *
+ * Note to future maintainers: This class should only contain methods which
+ * forward on to NameFormatter and nothing else; otherwise, it is no longer
+ * functioning in spirit as a mix-in.
+ */
+public abstract class NameFormatterMixin implements NameFormatter {
```

Comment:
> That view of design patterns is too narrow. There are cases of multiple design patterns having the same structure but having different intent - e.g. Strategy & Bridge: http://stackoverflow.com/questions/464524/what-is-the-difference-between-the-bridge-pattern-and-the-strategy-pattern . Thus, a design pattern is a design plus an intent. My intent is different from Proxy and Decorator, thus they don't fit.

I forgot about Strategy vs. Bridge. Thanks for reminding me. I think there's some nuance here between intent and solution, but we've spent enough time arguing philosophy for now. :)

> At a larger level, we shouldn't get to focused on design patterns. They are a tool that can be applied when appropriate, but we should never constrain our designs to fit the mold of design patterns when the intent does not match.

Agreed. When I was doing Scala, I barely thought about design patterns at all. I sometimes cynically think of design pattern as the contortions needed in OOP to do stuff that is natural in FP, but there are probably exceptions.

I only dug into design patterns here as a source of names to describe what you were trying to do.